### PR TITLE
ENH: stats.power: add function to simulate hypothesis test power

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
NA

#### What does this implement/fix?
[`statsmodels` ](https://www.statsmodels.org/devel/stats.html#power-and-sample-size-calculations) and [`pingouin`](https://pingouin-stats.org/build/html/api.html#power-analysis) offer analytical power calculation for a few of SciPy's hypothesis tests, but I thought it would be useful to offer a general function for simulating the power of essentially any SciPy hypothesis test. The same function can be used to study the effects of
- sample size(s),
- test parameters, and
- alternative hypothesis parameters (e.g. effect size)

on the power, and it can also be used to verify that p-values produced under the null hypothesis are uniformly distributed. In an enhancement, a method of the result object could be used to solve for an unknown (e.g. sample size, effect size) to achieve a given power, perhaps by (inverse) interpolation or a root finding method that is robust against noise.

[Documentation/examples](https://output.circle-artifacts.com/output/job/b12ecf2f-c374-420e-845f-e3a8d4e0bfd1/artifacts/0/html/reference/generated/scipy.stats.power.html#scipy.stats.power)

#### Additional information
I think this is ready to ready to review. Items not checked below could be left to follow-up work.
- [x] ~~improve auto-vectorization when callable does not have `axis` argument~~ (improved documentation)
- [x] improve the `args` interface (should it be `kwargs`? Do *all* `rvs` and `test` callables need to accept the `args`?)
- [x] simplify `n_observations`/`args` input validation (or at least review it... it was a bit rushed)
- [x] add tests
- [x] ~~generalize for other ways of assessing rejection of null hypothesis (e.g. observed value of statistic and critical value instead of p-value and significance threshhold)~~
- [ ] research / think about whether considerations like ["Permutation P-values should never be zero"](https://arxiv.org/pdf/1603.05766) apply here
- [ ] add `plot` method to result object?
- [ ] add `solve` method to result object?
- [x] support array-like `significance`
- [x] support `test` returning array-like p-value

### CODE REVIEW

**Comment by josef-pkt** [comment]
just one comment based on a brief look

When I check the null distribution (and power) of a hypothesis test, then I usually compare several p-value thresholds, `significance` (`alphas`).

It's a useful function.
I usually write it from scratch each time in a notebook. One reason is flexibility, very often I compute the rejection rate also for several methods. Many of the `stats` functions that I added to statsmodels, especially proportions and rates, have method options. In those cases I collect several p-values for each Monte Carlo replication.

**Comment by mdhaber** [comment]
Thanks for taking a look. I don't imagine this function will replace all ad hoc power simulations, but these suggestions will help make it cover more cases.

> When I check the null distribution (and power) of a hypothesis test, then I usually compare several p-value thresholds

Sure. It would be easy to allow `significance` to be an array. I considered it, but thought it might be enough to return all the p-values. Did you see the last example in the documentation in which I compare the ECDF of p-values against the uniform CDF?

> Many of the `stats` functions... have method options. In those cases I collect several p-values for each Monte Carlo replication.

It's possible to compute the power for multiple method options in a single call, but right now I'm only sure about doing it with independent Monte Carlo samples. I'll think about allowing the Monte Carlo samples to be shared - probably I'd just allow the user to return multiple p-values from the `test` callable.

**Comment by mdhaber** [comment]
@raphaelvallat do you have a moment to comment on the utility of such a function?

**Comment by raphaelvallat** [comment]
@mdhaber I think this would be a great addition to scipy 👍

File: scipy/stats/tests/test_resampling.py [hunk 1]
```diff
@@ -1010,6 +1010,172 @@ def statistic(*args, axis):
...
+        with pytest.raises(ValueError, match=message):
+            power(test, rvs, n_observations, batch=-10)
+        with pytest.raises(ValueError, match=message):
+            power(test, rvs, n_observations, batch=10.5)
+
+    def test_batch(self):
+        # make sure that the `batch` parameter is respected by checking the
+        # maximum batch size provided in calls to `test`
+        rng = np.random.default_rng(23492340193)
+        x = rng.random(10)
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:1074]:
```suggestion
```

File: scipy/stats/_resampling.py [hunk 2]
```diff
@@ -937,6 +937,302 @@ def two_sided(null_distribution, observed):
...
+    -------
+    res : PowerResult
+        An object with attributes:
+
+        power : float or ndarray
+            The estimated power against the alternative.
+        pvalues : ndarray
+            The p-values observed under the alternative hypothesis.
+
+    Examples
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:1114]:
```suggestion
            The p-values observed under the alternative hypothesis.

    Notes
    -----
    Suppose that `significance` is an array with shape ``shape1``, the elements
    of `kwargs` and `n_observations` are mutually broadcastable to shape ``shape2``,
    and `test` returns an array of p-values of shape ``shape3``. Then the result
    object ``power`` attribute will be of shape ``shape1 + shape2 + shape3``, and
    the ``pvalues`` attribute will be of shape ``shape2 + shape3 + (n_resamples,)``.
    
    Examples
```

File: scipy/stats/tests/test_resampling.py [hunk 3]
```diff
@@ -1010,6 +1010,172 @@ def statistic(*args, axis):
...
+        for significance_i in significance:
+            for nx_i in nx:
+                for alternative_i in kwargs['alternative']:
+                    for popmean_i in popmeans:
+                        def test(x, axis=-1):
+                            return stats.ttest_1samp(x, popmean_i, axis=axis,
+                                                     alternative=alternative_i)
+
+                        tmp = stats.power(test, rng.normal, nx_i,
+                                          significance=significance_i)
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:1138]:
```suggestion
                        def test2(x, axis=-1):
                            return stats.ttest_1samp(x, popmean_i, axis=axis,
                                                     alternative=alternative_i)

                        tmp = stats.power(test2, rng.normal, nx_i,
                                          significance=significance_i)
```

**Comment by fancidev** [comment]
In addition, it appears clearer to me if the arguments `(test, rvs, n_observations)` are renamed to `(test, alternative, sample_size)`, although I can see the current naming mirrors those of `monte_carlo_test`.

**Comment by josef-pkt** [comment]
just one remark

I think the function name should indicate simulation, e.g. power_simulated, simulate_power (montecarlo_power, ...)

**Comment by tupui** [comment]
> I think the function name should indicate simulation, e.g. power_simulated, simulate_power (montecarlo_power, ...)

I would not do that as it would add more jargon. The doc can explain the nuance for people who want to know what's behind the hood. I suspect that this won't matter for most users. So better keep it simple.

**Comment by josef-pkt** [comment]
Well, you use a pretty general term and use it for a special method.

power without simulations:
https://www.statsmodels.org/dev/stats.html#power-and-sample-size-calculations

But if scipy uses `power` term only for the special method, then I don't have to worry much that you steal my users.
:)

**Comment by tupui** [comment]
Why again this narrative about stealing users!? 🤷‍♂️ That's just not true and both projects definitely don't serve the same audience. Like Pingouin still being relevant even if this goes in.

**Comment by fancidev** [comment]
It’s probably most important to define the signature at this point as the implementation can always be refined later. To choose the “nicest” signature it may be helpful to postulate a few example use cases and see if the signature leads to clear and easily understandable code.

Some uses cases are:

- one sample t test that the mean of X is equal to mu.
- Two sample t test that the mean of X and Y are equal.

**Comment by tupui** [comment]
I personally prefer Matt's proposal with passing the significance as a parameter (also vectorized to allow exploration). I am not sure what we would facilitate by having a method like you propose instead. What is your rationale here?

For kwargs I agree that it can be confusing and would prefer not to propose it first. Though I get that it is sort of needed if we want to vectirize on these. I am just wondering if people would really use that capability. Maybe for later if there is a demand for it.

**Comment by fancidev** [comment]
The last example in the doc, which uses the `power` function to simulate the [size](https://en.m.wikipedia.org/wiki/Size_(statistics)) of a test, feels like an anti-pattern to me and probably should not be advertised. It is like including an example for the `sin` function to say it can be used to compute cosine by shifting the argument by `0.5*pi`.

But this “duality” does suggest that the function may be able to be generalized a bit; but as always, the right level of generality is hard to design!

**Comment by tupui** [comment]
I am not sure I understand what you suggest then @fancidev

**Comment by tupui** [comment]
> I just wanted to point out that I don't think of simulations when reading power. Using the term for simulations will limit what other "powers" there might be.

That's a fair point. The question to me is more if we were to add more strategies to compute the power, would we want to add new functions or just use a parameter to set the method. I don't expect a lot of users to tweak that. For power user, when that time comes, we can still have separate functions and keep a higher level API.

**Comment by josef-pkt** [comment]
@fancidev 

Checking the size is important, but for users not because of checking whether it's "implemented correctly"

I would add an example where a hypothesis test has not the right size.

for example, because

- some underlying assumption does not hold, e.g. use equal-var t-test when variances in two samples differ. (unequal var Welch test would be correctly sized)
- discreteness of the sample space like binomial counts, test cannot have size=alpha but could maintain size <= alpha. Fisher's  exact test maintains size but is conservative, size can be much smaller than alpha on average.
- asymptotic, large sample distribution and p-values are not a good approximation for small samples. e.g. AFAIR, many variance/dispersion tests can have incorrect size in small samples, especially if kurtosis is larger than in normal distribution.

**Comment by mdhaber** [comment]
> the doc on the return value of `stats.power` could be more elaborate

I will probably add more information in the Notes. 

> If that’s the case, would it be more handy/logical

"handy", perhaps; "*more* logical" is hard to judge. In this case, I followed the precedent set by `bootstrap`, which returns the confidence interval as an array when it is arguable that it should be a method. I've regretted that choice sometimes, but I don't recall any reported complaints.

> Personally I’m interested to see for example a K-S goodness of fit test of normal null vs say t alternative.

If the purpose is to show the very low power of K-S compared to other normality tests, that sounds like a nice addition. We'll see whether the reviewing maintainer thinks it's essential or if the first example is simple enough for them. 

> It seems just easier to let the caller use a lambda if they want to achieve this.

I have spoken against the inclusion of `args`/`kwargs` parameters in the past (e.g. see e.g. gh-13258). I agree that they are almost always redundant and *usually* have little added value, so normally I'm against adding them, too. (Others have disagreed and recommended inclusion.) Certainly it is easier for us to leave them out, but adding support for them is a common feature request (e.g. gh-8352 or many Stack Overflow questions).

It has been several months since I wrote this, so I can't be sure what pushed me in the other direction here, but it was probably that I could think of very few uses of this function *without* the user having to write a wrapper. I think I remember considering that the examples looked a more complicated without it, and reviewers would not like that, either. 

There is a balance to be struck here.

> The last example in the doc, which uses the `power` function to simulate the size of a test, feels like an anti-pattern and probably should not be advertised.

The difference is that the `cos` function exists.

> I would add an example where a hypothesis test has not the right size.

That's a good idea. I like that better than suggesting that our tests might not be implemented correctly (although I have caught problems in PRs by simulating their size). I can add that. Is that what you meant in https://github.com/scipy/scipy/pull/19561#issuecomment-1992867265?

(Note that the accuracy of the distribution of the test statistic for small samples is already discussed extensively in the [Resampling and Monte Carlo Methods](https://docs.scipy.org/doc/scipy/tutorial/stats/resampling.html) tutorials and the API documentation of several tests based on asymptotic approximations (e.g. `normaltest`) and resampling/MC methods (e.g. `monte_carlo_test`) . 

Also, if you wanted to this to compare against the power result produce by one of the `statsmodels` power calculations, please suggest an example. I'd be happy to include it (if we can use `statsmodels` in examples).

**Review by tupui — APPROVED** [review]
Approving now as I think the API is sensible and there is nothing that couldn't be done in a follow up. Unless there is something you really want to have in this PR, I would propose to move forward and merge.

File: scipy/stats/_resampling.py [hunk 4]
```diff
@@ -1,5 +1,3 @@
-from __future__ import annotations
-
 import warnings
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:1]:
```suggestion
from __future__ import annotations

import warnings
```

**Comment by tupui** [comment]
No new comments and the new commits address others suggestions regarding the doc. Let's get this in! Thanks Matt and everyone else for reviewing 🙌

**Pull request merged**